virtual_ship/sailship.py

```diff
@@ -222,68 +86,348 @@ def sailship(config: VirtualShipConfig):
+    while not cruise.finished:
+        measurements.adcps.append(cruise.spacetime)
+        yield _WaitFor(sample_period)
+
+
+def _ctd_task(
+    stationkeeping_time: timedelta,
+    min_depth: float,
+    max_depth: float,
+    cruise: _Cruise,
+    measurements: _MeasurementsToSimulate,
+) -> Generator[_WaitFor, None, None]:
+    with cruise.do_not_sail():
+        measurements.ctds.append(
+            CTD(
+                spacetime=cruise.spacetime,
+                min_depth=min_depth,
+                max_depth=max_depth,
+            )
         )
+        yield _WaitFor(stationkeeping_time)
+
+
+def _drifter_task(
+    cruise: _Cruise, measurements: _MeasurementsToSimulate, config: VirtualShipConfig
+) -> None:
+    measurements.drifters.append(
+        Drifter(
+            cruise.spacetime,
+            depth=config.drifter_config.depth,
+            lifetime=config.drifter_config.lifetime,
+        )
+    )
+
 
-    # explitly include final point which is not added by the previous loop
-    sample_points.append(
-        Location(
-            latitude=config.route_coordinates[-1].lat,
-            longitude=config.route_coordinates[-1].lon,
+def _argo_float_task(
+    cruise: _Cruise, measurements: _MeasurementsToSimulate, config: VirtualShipConfig
+) -> None:
+    measurements.argo_floats.append(
+        ArgoFloat(
+            spacetime=cruise.spacetime,
+            min_depth=config.argo_float_config.min_depth,
+            max_depth=config.argo_float_config.max_depth,
+            drift_depth=config.argo_float_config.drift_depth,
+            vertical_speed=config.argo_float_config.vertical_speed,
+            cycle_days=config.argo_float_config.cycle_days,
+            drift_days=config.argo_float_config.drift_days,
         )
     )
 
-    return sample_points
+
+def _verify_waypoints(
+    waypoints: list[Waypoint], ship_speed: float, projection: pyproj.Geod
+) -> None:
+    """
+    Verify waypoints are ordered by time, first waypoint has a start time, and that schedule is feasible in terms of time if no unexpected things happen.
+
+    :param waypoints: The waypoints to check.
+    :param ship_speed: Speed of the ship.
+    :param projection: projection used to sail between waypoints.
+    :raises PlanningError: If waypoints are not feasible or incorrect.
+    """
+    # check first waypoint has a time
+    if waypoints[0].time is None:
+        raise PlanningError("First waypoint must have a specified time.")
+
+    # check waypoint times are in ascending order
+    timed_waypoints = [wp for wp in waypoints if wp.time is not None]
+    if not all(
+        [
+            next.time >= cur.time
+            for cur, next in zip(timed_waypoints, timed_waypoints[1:])
+        ]
+    ):
+        raise PlanningError(
+            "Each waypoint should be timed after all previous waypoints"
+        )
+
+    # check that ship will arrive on time at each waypoint (in case nothing goes wrong)
```

Comment:
```suggestion
    # check that ship will arrive on time at each waypoint (in case no unexpected events happen)
```

Reply:
Let's call these 'unexpected events' throughout, so we can later add storms etc if we want